### ISSUE TITLE
[PythonDispatch] when the value of a argument is the same as its default value, __torch_dispatch fail to capture it.

### ISSUE THREAD

**Issue by LamForest** [pytorch/pytorch#121917]
## Background
We are using `__torch_dispatch__` to capture arguments passed to aten ops. Then we analyze or dump arguments to facilitate model training.

## Issue
When a default value is defined for a op in native_functions.yaml and passed argument **is same as default value**, such arguments will be ignored.

## repro
```py
from torch.utils._python_dispatch import TorchDispatchMode, _pop_mode_temporarily
import torch

class LoggingArgs(TorchDispatchMode):
    def __torch_dispatch__(self, op, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        print(f"{args=}")
        print(f"{kwargs=}")
        return op(*args, **kwargs)

x = torch.randn(3)
with LoggingArgs():
    print("-----keepdim = False-------") 
    x.sum(0, keepdim=False) #default value is False

    print("-----keepdim = True-------")
    x.sum(0, keepdim=True)
    
    print("-----negative_slope=0.01-------")
    torch.nn.functional.leaky_relu(x, negative_slope=0.01) #default value is 0.01
    
    print("-----negative_slope=1.0-------")
    torch.nn.functional.leaky_relu(x, negative_slope=1.0)
```

output:
```
-----keepdim = False-------
args=(tensor([-0.0603,  0.3735, -1.5986]), [0]) #keepdim = False is not captured
kwargs={}
-----keepdim = True-------
args=(tensor([-0.0603,  0.3735, -1.5986]), [0], True)
kwargs={}
-----negative_slope=0.01-------
args=(tensor([-0.0603,  0.3735, -1.5986]),) #negative_slope = 0.01 is not captured
kwargs={}
-----negative_slope=1.0-------
args=(tensor([-0.0603,  0.3735, -1.5986]), 1.0)
kwargs={}
```
I can reproduce this on 2.0.1 and latest main.

### Versions

```
Collecting environment information...
PyTorch version: 2.3.0a0+gitb0aa01b
Is debug build: True
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.22.2
Libc version: glibc-2.31

Python version: 3.11.8 (main, Feb 26 2024, 21:39:34) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.14.0_1-0-0-48-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          128
On-line CPU(s) list:             0-127
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Platinum 8350C CPU @ 2.60GHz
Stepping:                        6
Frequency boost:                 enabled
CPU MHz:                         800.000
CPU max MHz:                     3500.0000
CPU min MHz:                     800.0000
BogoMIPS:                        5200.00
Virtualization:                  VT-x
L1d cache:                       3 MiB
L1i cache:                       2 MiB
L2 cache:                        80 MiB
L3 cache:                        96 MiB
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts avx512vbmi umip pku avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.8.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] optree==0.10.0
[pip3] torch==2.3.0a0+gitb0aa01b
[conda] blas                      1.0                         mkl    defaults
[conda] mkl                       2023.1.0         h213fc3f_46344    defaults
[conda] mkl-fft                   1.3.8                    pypi_0    pypi
[conda] mkl-random                1.2.4                    pypi_0    pypi
[conda] mkl-service               2.4.0                    pypi_0    pypi
[conda] mkl_fft                   1.3.8           py311h5eee18b_0    defaults
[conda] mkl_random                1.2.4           py311hdb19cb5_0    defaults
[conda] numpy                     1.26.0                   pypi_0    pypi
[conda] numpy-base                1.26.4          py311hf175353_0    defaults
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] torch                     2.3.0a0+gitb0aa01b           dev_0    <develop>
[conda] torchfix                  0.4.0                    pypi_0    pypi
```

**Comment by LamForest** [comment]
cc @ezyang @albanD @Chillee @zou3519

**Comment by albanD** [comment]
This is the expected behavior.
The expectation is that you can access the default value from `func`'s schema if you need it.
Would this work for your use case?

**Comment by LamForest** [comment]
> This is the expected behavior. The expectation is that you can access the default value from `func`'s schema if you need it. Would this work for your use case?

Thanks for your reply. @albanD 

I am afraid not. In our use case, if users call `x.sum(0, keepdim=False)`, we want to dump aten op's arguments `{ "self": [xx, xx], "dim": 0, "keepdim" : False}`. Otherwise, users would be confused why keepdim is missing.

As you said, I can access default value through schema, but, I an not able to know which argument's default value I need to access, because I don’t know which parameter has been missed.

The only way is to always dump all arguments. However, some ops has many arguments with default value. For example, we have a custom fused op looks like this:
```
(..., int head_dim, float softmax_scale=0.0, float p_dropout=0.0, int seed=0x0, bool is_causal=True, Tensor? attn_mask=None, Tensor? bias=None, Tensor? q_max=None, Tensor? k_max=None, Tensor? v_max=None, Tensor? o_max=None) -> (Tensor(b!), Tensor(c!))
```
If dumpling all arguments, the dumping would be too long and contain unimportant info.

**Comment by ezyang** [comment]
It's not great because you have to hard code in the defaults, but I like this helper function https://github.com/pytorch/pytorch/pull/117557/files#diff-c906a4eb35b04bec3fa52b8f64fa9707b7e91da4736c516fabf469e455938a1bR21-R43
